lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,273 @@
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.logsafe.Preconditions;
+
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final HashMap<Long, MapEntry> timestampMap = new HashMap<>();
+    private final TreeMultimap<IdentifiedVersion, Long> aliveVersions =
+            TreeMultimap.create(IdentifiedVersion.comparator(), Ordering.natural());
+
+    @GuardedBy("this")
+    private boolean failed = false;
+
+
+    private LockWatchEventCacheImpl(ClientLockWatchEventLog eventLog) {
+        this.eventLog = eventLog;
+    }
+
+    public static LockWatchEventCacheImpl create() {
+        return create(ClientLockWatchEventLogImpl.create());
+    }
+
+    @VisibleForTesting
+    static LockWatchEventCacheImpl create(ClientLockWatchEventLog eventLog) {
+        return new LockWatchEventCacheImpl(eventLog);
+    }
+
+    @Override
+    public Optional<IdentifiedVersion> lastKnownVersion() {
+        return eventLog.getLatestKnownVersion();
+    }
+
+    @Override
+    public synchronized void processStartTransactionsUpdate(
+            Set<Long> startTimestamps,
+            LockWatchStateUpdate update) {
+        ensureNotFailed(() -> {
+            Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+            latestVersion.ifPresent(
+                    version -> startTimestamps.forEach(timestamp -> {
+                        timestampMap.put(timestamp, MapEntry.of(version));
+                        aliveVersions.put(version, timestamp);
+                    }));
+        });
+    }
+
+    @Override
+    public synchronized void processGetCommitTimestampsUpdate(
+            Collection<TransactionUpdate> transactionUpdates,
+            LockWatchStateUpdate update) {
+        ensureNotFailed(() -> {
+            Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+            latestVersion.ifPresent(version ->
+                    transactionUpdates.forEach(transactionUpdate -> {
+                        MapEntry previousEntry = timestampMap.get(transactionUpdate.startTs());
+                        Preconditions.checkNotNull(previousEntry, "Start timestamp missing from cache");
+                        timestampMap.replace(transactionUpdate.startTs(), previousEntry.withCommitInfo(
+                                CommitInfo.of(transactionUpdate.commitTs(),
+                                        transactionUpdate.writesToken(),
+                                        version)));
+                    }));
+        });
+    }
+
+    @Override
+    public synchronized CommitUpdate getCommitUpdate(long startTs) {
+        checkNotFailed();
+        MapEntry entry = timestampMap.get(startTs);
+        Preconditions.checkState(entry.commitInfo().isPresent(), "Commit timestamp update not yet processed");
```

Comment:
Is this the only case here? I feel like there's a legitimate case of "leader switched", in which case you should throw an exception that will cause the transaction to retry.

Please review all places that you throw here and probably introduce some custom exceptions that extend TransactionFailedRetriableException.

Reply:
Oh wait, this one is fine.

Reply:
I mentioned this in a comment today - I think throwing a retriable exception sounds reasonable.